### ISSUE TITLE
Create a mechanism to trigger tests that only warn once

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#47624]
Today in PyTorch we have [TORCH_WARN](https://github.com/pytorch/pytorch/blob/c9caa828f5320fe635b6e785cc007654a4177a1b/c10/util/Exception.h#L390) and [TORCH_WARN_ONCE](https://github.com/pytorch/pytorch/blob/c9caa828f5320fe635b6e785cc007654a4177a1b/c10/util/Exception.h#L396). TORCH_WARN_ONCE, as its name suggests, only throws a warning one time. This makes it tricky to test for. Our test suite, for example, has [maybeWarnsRegex](https://github.com/pytorch/pytorch/blob/4a2fb3404215c774fdf2805ba8d495bb7dc1bbbf/torch/testing/_internal/common_utils.py#L1196), but this just verifies that no other warnings are thrown.

This feature request is for a mechanism to consistently trigger warnings that use TORCH_WARN_ONCE. One mechanism for this would be to add a new context that, if set, causes these warnings to always warn. This would let us reliably test that these warnings are thrown correctly. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @mruberry @VitalyFedyunin @walterddr

**Comment by mattip** [comment]
If I understand the code correctly, it seems `TORCH_WARN_ONCE` and `TORCH_WARN` do not currently always use python. In a call like [setDeterministic](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Context.cpp#L66), the default `warning_handler` is used, which goes to [`LOG_AT_FILE_LINE`](https://github.com/pytorch/pytorch/blob/master/c10/util/Exception.cpp#L128) which uses [`MessgeLogger`](https://github.com/pytorch/pytorch/blob/master/c10/util/logging_is_not_google_glog.h#L110) to  [print to std::err](https://github.com/pytorch/pytorch/blob/master/c10/util/Logging.cpp#L304). Most python wrappers use the [`HANDLE_TH_ERRORS` macro](https://github.com/pytorch/pytorch/blob/master/torch/csrc/Exceptions.h#L45) to set a `warning_handler` by calling [PyWarningHandler](https://github.com/pytorch/pytorch/blob/master/torch/csrc/Exceptions.cpp#L176). For some reason [the wrapper for `setDeterministic`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/Module.cpp#L455) does not use `HANDLE_TH_ERRORS`.

But that is only one part of the story. Overriding the behaviour of `TORCH_WARN_ONCE` to make it warn on each call is more complicated. As defined in [`c10/util/Exception.h`](https://github.com/pytorch/pytorch/blob/master/c10/util/Exception.h#L401), it instantiates a static class instance, making it a singleton, and uses that instance to call  [`c10::Warning::warn`](https://github.com/pytorch/pytorch/blob/master/c10/util/Exception.cpp#L111) only once ever. Refactoring that to call each time would make this equivalent to `TORCH_WARN`.

**Comment by mruberry** [comment]
I was hoping we could do something like set a global context, "setAlwaysWarn(true/false)", that TORCH_WARN_ONCE would query and, if the global flag was true, it would then warn like TORCH_WARN does.

**Comment by mattip** [comment]
Is this the direction you were thinking?
```
#define TORCH_WARN_MAYBE(...) \
  if (g_WARN_ONLY_ONCE) { \
    TORCH_WARN_ONCE(...) \
  } else { \
    TORCH_WARN(...) \
  }
```
Then 
- Some/All of the uses of `TORCH_WARN_ONCE` would be refactored to `TORCH_WARN_MAYBE`. 
- The new boolean global value `g_WARN_ONLY_ONCE` would be set `false` by the python warning handler, passing responsibility for filtering warnings to python.
- Tests would need a warning filter, and could check that the warning is emitted.
 
Is there a warning context manager in the testing framework, or should we use the [`numpy.testing.suppress_warnings`](https://numpy.org/devdocs/reference/generated/numpy.testing.suppress_warnings.html) context manager to work around [python issues with warning filters](https://bugs.python.org/issue29672)

**Comment by mruberry** [comment]
That's close to what I was thinking of.

I was thinking we'd keep TORCH_WARN_ONCE, and there would be a flag, like `g_WARN_ONLY_ONCE`, that had the behavior you're describing. It's true that this would make TORCH_WARN_ONCE a bit of a misnomer, but that's OK.

I was thinking the flag would be controlled by a PyTorch global context (like the deterministic flag). 

We currently use a the Python warning filter in our test suite and that seems to be OK, I don't think we'd need to use numpy.testing.suppress_warnings. 

Basically we want this flag to be just for debugging. We'll set it to true during the test suite so we can verify that these warnings are thrown.

**Comment by mattip** [comment]
Now that the `assertWarnsOnceRegex` is inplace, the next step is to replace `maybeWarnsRegex` with it.

**Comment by mruberry** [comment]
I think this issue is closed. @mattip, would you file a separate issue to audit uses of maybeWarnsRegex in the test suite?

**Comment by mattip** [comment]
See issue #52025
